### PR TITLE
Fix `MRMRFeatureSelectionTransform`: change `less_is_better` logic, add `drop_zero` mode

### LINKED ISSUE #308: Fix MRMR: change `less_is_better` logic, add `drop_zero` mode
### 🚀 Feature Request

1. Add a check that all elements of the relevance table are greater than or equal to zero
2.  If mode is less_is_better we change relevance of features using this logic:
$feature_{i}.relevance =  \max{feature.relevance} + \min{feature.relevance} - feature_{i}.relevance$

4. Change formula for score from
$feature_{i}.score = \frac{feature_{i}.relevance}{mean(corr(feature_{i},  feature_{in} ))}$
to $feature_{i}.score = feature_{i}.relevance(1 - max(corr(feature_{i},  feature_{in} )))$.
$feature_{in}$ is designation of already selected feature.

4.Add drop_zero mode(mode there feature with zero relevance will not be used in selection) 
If $top_k \geq count(nonzerofeatures)$ MRMR will choose all all features with relevance > 0 and randomly add $top_k - count(nonzerofeatures)$ features with relevance = 0.

### Proposal

Fix MRMR logic and increase it speed in some cases 

### Test cases

1.Test Cases that MRMRFeatureSelectionTransform raises Error if there are incorrect relevance
2.Test Cases for score
4.Test cases for drop_zero mode

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by yellowssnake**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

Fix MRMRFeatureSelectionTransform, change score formula and add drop_zero mode

## Closing issues
Closes #308.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -42,835 +42,5 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - 
 
 ### Fixed
-- Fix FordA download url in classification notebook ([#309](https://github.com/etna-team/etna/pull/309))
```
Comment (Ama16) [inline CHANGELOG.md:45]: What is it?...

File: etna/analysis/feature_selection/mrmr_selection.py [hunk 2]
```diff
@@ -59,6 +71,10 @@ def mrmr(
     fast_redundancy:
         * True: compute redundancy only inside the the segments, time complexity :math:`O(top\_k * n\_segments * n\_features * history\_len)`
         * False: compute redundancy for all the pairs of segments, time complexity :math:`O(top\_k * n\_segments^2 * n\_features * history\_len)`
+    drop_zero:
+        If True, drop features with zero relevance before MRMR. If top_k is greater number of features
```
Comment (Ama16) [inline etna/analysis/feature_selection/mrmr_selection.py:75]: It is not clear for me. If top_k is greater than number of features with relevance >0, select all non-zero relevance features with randomly selected zero relevance features.

File: etna/analysis/feature_selection/mrmr_selection.py [hunk 3]
```diff
@@ -88,12 +104,25 @@ def mrmr(
...
     not_selected_features = all_features.copy()
 
-    redundancy_table = pd.DataFrame(np.inf, index=all_features, columns=all_features)
-    top_k = min(top_k, len(all_features))
+    if drop_zero is True:
+        not_relevant_features = list(
+            filter(lambda feature: is_not_relevant(relevance, feature, atol), not_selected_features)
+        )
+        not_selected_features = list(
+            filter(lambda feature: is_relevant(relevance, feature, atol), not_selected_features)
```
Comment (Ama16) [inline etna/analysis/feature_selection/mrmr_selection.py:120]: Here you compute twice the same logic. Could you do it in one compute?

File: etna/analysis/feature_selection/mrmr_selection.py [hunk 4]
```diff
@@ -88,12 +104,25 @@ def mrmr(
...
+
     segments = set(regressors.columns.get_level_values("segment"))
     selected_features: List[str] = []
     not_selected_features = all_features.copy()
 
-    redundancy_table = pd.DataFrame(np.inf, index=all_features, columns=all_features)
-    top_k = min(top_k, len(all_features))
+    if drop_zero is True:
+        not_relevant_features = list(
+            filter(lambda feature: is_not_relevant(relevance, feature, atol), not_selected_features)
```
Comment (Ama16) [inline etna/analysis/feature_selection/mrmr_selection.py:117]: Why we compare with atol? Why not strong equality to 0?

File: etna/transforms/feature_selection/feature_importance.py [hunk 5]
```diff
@@ -235,8 +235,14 @@ def _fit(self, df: pd.DataFrame) -> "MRMRFeatureSelectionTransform":
         df_features = df_features.loc[df_features.first_valid_index() :]
         relevance_table = self.relevance_table(df_target, df_features, **self.relevance_params)
 
+        if relevance_table.values.min() < 0:
```
Comment (Ama16) [inline etna/transforms/feature_selection/feature_importance.py:238]: you should add it into relevance_table, not here
Here you should have guarantees that all values >= 0

File: etna/analysis/feature_selection/mrmr_selection.py [hunk 6]
```diff
@@ -30,11 +31,22 @@ class AggregationMode(str, Enum):
...
+    """Return False if feature is relevant, else return True."""
+    return math.isclose(relevance.loc[feature], 0, abs_tol=atol)
+
+
 def mrmr(
     relevance_table: pd.DataFrame,
     regressors: pd.DataFrame,
     top_k: int,
     fast_redundancy: bool = False,
+    drop_zero: bool = True,
```
Comment (Ama16) [inline etna/analysis/feature_selection/mrmr_selection.py:49]: You cant use this parameter in transform. You should add it in init

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py [hunk 7]
```diff
@@ -319,6 +319,99 @@ def test_mrmr_right_regressors(relevance_table, ts_with_regressors, fast_redunda
...
+        model=CatBoostRegressor(iterations=1),
+        fast_redundancy=fast_redundancy,
+    )
+    df_selected = mrmr.fit_transform(ts).to_pandas()
+    selected_regressors = set()
+    for column in df_selected.columns.get_level_values("feature"):
+        if column.startswith("regressor"):
+            selected_regressors.add(column)
+    assert selected_regressors == {
+        "regressor_useless_11",
```
Comment (Ama16) [inline tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:340]: Why method should save specifically this useless regressor?

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py [hunk 8]
```diff
@@ -319,6 +319,99 @@ def test_mrmr_right_regressors(relevance_table, ts_with_regressors, fast_redunda
...
+        "regressor_useless_8",
+        "regressor_useless_1",
+        "regressor_useless_2",
+        "regressor_useless_0",
+    }
+
+
+@pytest.mark.parametrize("fast_redundancy", ([True, False]))
+@pytest.mark.parametrize("relevance_table", ([ModelRelevanceTable()]))
+def test_mrmr_drop_zero_mode_sanity_check(relevance_table, ts_with_regressors, fast_redundancy):
```
Comment (Ama16) [inline tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:360]: Where is the drop zero mode tests here?

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py (hunk 7 above)
Comment (Ama16) [reply tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:340]: if here top_k>regressors, than make assert on equal length, not exact names

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py [hunk 9]
```diff
@@ -319,6 +319,99 @@ def test_mrmr_right_regressors(relevance_table, ts_with_regressors, fast_redunda
...
+    for column in df_selected.columns.get_level_values("feature"):
+        if column.startswith("regressor"):
+            selected_regressors.add(column)
+
+    assert set(selected_regressors) == {"regressor_useful_0", "regressor_useful_1", "regressor_useful_2"}
+
+
+@pytest.mark.parametrize("fast_redundancy", ([True, False]))
+@pytest.mark.parametrize("relevance_table", ([ModelRelevanceTable()]))
+def test_mrmr_select_top_k_regressors_in_drop_zero_mode(relevance_table, ts_with_regressors, fast_redundancy):
```
Comment (Ama16) [inline tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:378]: where is drop zero mode tests here x2?

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py [hunk 10]
```diff
@@ -319,6 +319,99 @@ def test_mrmr_right_regressors(relevance_table, ts_with_regressors, fast_redunda
...
+    selected_regressors = set()
+    for column in df_selected.columns.get_level_values("feature"):
+        if column.startswith("regressor"):
+            selected_regressors.add(column)
+    assert len(selected_regressors) == 10
+
+
+@pytest.mark.parametrize("fast_redundancy", ([True, False]))
+@pytest.mark.parametrize("relevance_table", ([ModelRelevanceTable()]))
+def test_mrmr_drop_zero_mode_sanity_check(relevance_table, ts_with_regressors, fast_redundancy):
```
Comment (Ama16) [inline tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:398]: You already have test with this name. And with meaning too

File: tests/test_transforms/test_inference/test_inverse_transform.py [hunk 11]
```diff
@@ -221,14 +221,14 @@ def _test_inverse_transform_train(self, ts, transform, expected_changes):
                     fast_redundancy=True,
                 ),
                 "ts_with_exog",
-                {"create": {"weekday", "monthday", "positive"}},
+                {"create": {"positive", "weekday", "year"}},
```
Comment (Ama16) [inline tests/test_transforms/test_inference/test_inverse_transform.py:224]: Why?

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py [hunk 12]
```diff
@@ -319,6 +319,99 @@ def test_mrmr_right_regressors(relevance_table, ts_with_regressors, fast_redunda
...
+
+    df_selected = mrmr.fit_transform(ts).to_pandas()
+    selected_regressors = set()
+    for column in df_selected.columns.get_level_values("feature"):
+        if column.startswith("regressor"):
+            selected_regressors.add(column)
+
+    assert set(selected_regressors) == {"regressor_useful_0", "regressor_useful_1", "regressor_useful_2"}
+
+
```
Comment (Ama16) [inline tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:414]: You added 0 tests on drop_zero mode. Please, add it

**Review by Ama16 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (Ama16) [reply CHANGELOG.md:45]: You delete all changelog

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py (hunk 10 above)
Comment (yellowssnake) [reply tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:398]: In this test I check that I select correct regressors in previous test I check that if top_k >= number of features with non-zero relevance it still select top_k

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py (hunk 9 above)
Comment (yellowssnake) [reply tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:378]: Answer in the next comment

File: etna/transforms/feature_selection/feature_importance.py (hunk 5 above)
Comment (yellowssnake) [reply etna/transforms/feature_selection/feature_importance.py:238]: Why? We can use relevance table for some other things, and only in MRMR we need non negative relevance

File: etna/analysis/feature_selection/mrmr_selection.py (hunk 2 above)
Comment (yellowssnake) [reply etna/analysis/feature_selection/mrmr_selection.py:75]: I select all non-zero relevance features and add some features with zero relevance in order to select exact top_k features

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py (hunk 8 above)
Comment (yellowssnake) [reply tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:360]: I add test where top_k > number of regressors, test where number of regressors with non-zero relevance <top_k<number of regressors and test to check sanity. I think this covers all the main cases. Maybe I should rename some tests

File: tests/test_transforms/test_inference/test_inverse_transform.py (hunk 11 above)
Comment (yellowssnake) [reply tests/test_transforms/test_inference/test_inverse_transform.py:224]: Because selected features changes

File: etna/analysis/feature_selection/mrmr_selection.py [hunk 13]
```diff
@@ -30,11 +30,22 @@ class AggregationMode(str, Enum):
 }
 
 
+def is_relevant(relevance, feature, atol):
```
Comment (Ama16) [inline etna/analysis/feature_selection/mrmr_selection.py:33]: Atol?

File: etna/analysis/feature_selection/mrmr_selection.py [hunk 14]
```diff
@@ -30,11 +30,22 @@ class AggregationMode(str, Enum):
...
+    """Return True if feature is relevant, else return False."""
+    return not relevance.loc[feature] == 0
+
+
+def is_not_relevant(relevance, feature, atol):
```
Comment (Ama16) [inline etna/analysis/feature_selection/mrmr_selection.py:38]: Do you really need this functions? You could just add this logic into filter explicitly without functions

File: etna/analysis/feature_selection/mrmr_selection.py (hunk 2 above)
Comment (Ama16) [reply etna/analysis/feature_selection/mrmr_selection.py:74]: I mean what is written is very unclear and complicated. Could you rewrite more clear?

File: etna/transforms/feature_selection/feature_importance.py [hunk 15]
```diff
@@ -172,6 +172,7 @@ def __init__(
         top_k: int,
         features_to_use: Union[List[str], Literal["all"]] = "all",
         fast_redundancy: bool = False,
+        drop_zero: bool = False,
```
Comment (Ama16) [inline etna/transforms/feature_selection/feature_importance.py:175]: Docstring?

File: etna/transforms/feature_selection/feature_importance.py (hunk 5 above)
Comment (Ama16) [reply etna/transforms/feature_selection/feature_importance.py:240]: From any relevance table the importance must be non-negative just because it is relevance

File: tests/test_analysis/test_feature_selection/test_mrmr.py [hunk 16]
```diff
@@ -83,25 +83,6 @@ def test_mrmr_right_regressors(df_with_regressors, relevance_method, expected_re
     assert set(selected_regressors) == set(expected_regressors)
 
 
-@pytest.mark.parametrize("fast_redundancy", [True, False])
```
Comment (Ama16) [inline tests/test_analysis/test_feature_selection/test_mrmr.py:86]: why?

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py [hunk 17]
```diff
@@ -319,6 +319,71 @@ def test_mrmr_right_regressors(relevance_table, ts_with_regressors, fast_redunda
...
+        model=CatBoostRegressor(iterations=1),
+        drop_zero=True,
+        fast_redundancy=fast_redundancy,
+    )
+    df_selected = mrmr.fit_transform(ts).to_pandas()
+    selected_regressors = set()
+    for column in df_selected.columns.get_level_values("feature"):
+        if column.startswith("regressor"):
+            selected_regressors.add(column)
+    assert len(selected_regressors) == 15
```
Comment (Ama16) [inline tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:340]: 15 is a magic constant. Take it from ts_with_regressors

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py [hunk 18]
```diff
@@ -319,6 +319,71 @@ def test_mrmr_right_regressors(relevance_table, ts_with_regressors, fast_redunda
...
+
+
+@pytest.mark.parametrize("fast_redundancy", ([True, False]))
+@pytest.mark.parametrize("relevance_table", ([ModelRelevanceTable()]))
+def test_mrmr_select_top_k_regressors_in_drop_zero_mode(relevance_table, ts_with_regressors, fast_redundancy):
+    """Check that transform selects right top_k regressors."""
```
Comment (Ama16) [inline tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:346]: I think its wrong description

**Review by Ama16 — CHANGES_REQUESTED** [review]

File: etna/analysis/feature_selection/mrmr_selection.py [hunk 19]
```diff
@@ -35,41 +35,46 @@ def mrmr(
...
-    redundancy_aggregation_mode:
-        the method for redundancy values per-segment aggregation
-    atol:
-        the absolute tolerance to compare the float values
-
-    Returns
-    -------
-    selected_features: List[str]
-        list of ``top_k`` selected regressors, sorted by their importance
+      Maximum Relevance and Minimum Redundancy feature selection method.
```
Comment (Ama16) [inline etna/analysis/feature_selection/mrmr_selection.py:44]: Why you add spaces here?

File: etna/analysis/feature_selection/mrmr_selection.py [hunk 20]
```diff
@@ -35,41 +35,46 @@ def mrmr(
...
+          contains relevance of j-th ``df_exog`` series to i-th df series
+      regressors:
+          dataframe with regressors in etna format
+      top_k:
+          num of regressors to select; if there are not enough regressors, then all will be selected
+      fast_redundancy:
+          * True: compute redundancy only inside the the segments, time complexity :math:`O(top\_k * n\_segments * n\_features * history\_len)`
+          * False: compute redundancy for all the pairs of segments, time complexity :math:`O(top\_k * n\_segments^2 * n\_features * history\_len)`
+    drop_zero:
+          * True: use only features with relevance > 0 in calculations, if their number is less than zero
```
Comment (Ama16) [inline etna/analysis/feature_selection/mrmr_selection.py:64]: `less than top_k` maybe?

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py [hunk 21]
```diff
@@ -319,6 +319,72 @@ def test_mrmr_right_regressors(relevance_table, ts_with_regressors, fast_redunda
...
+        model=CatBoostRegressor(iterations=1),
+        drop_zero=True,
+        fast_redundancy=fast_redundancy,
+    )
+    df_selected = mrmr.fit_transform(ts).to_pandas()
+    selected_regressors = set()
+    for column in df_selected.columns.get_level_values("feature"):
+        if column.startswith("regressor"):
+            selected_regressors.add(column)
+    assert len(selected_regressors) == 10
```
Comment (Ama16) [inline tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:362]: If you know all relevanced features, add a check for their presence

**Review by Ama16 — DISMISSED** [review]

File: tests/test_transforms/test_feature_selection/test_feature_importance_transform.py [hunk 22]
```diff
@@ -319,6 +319,95 @@ def test_mrmr_right_regressors(relevance_table, ts_with_regressors, fast_redunda
...
+@pytest.mark.parametrize("relevance_table", ([ModelRelevanceTable()]))
+def test_mrmr_top_k_greater_than_number_of_regressors(relevance_table, ts_with_regressors, fast_redundancy):
+    """Check that transform selects all regressors if top_k greater than number of regressors."""
+    ts = ts_with_regressors
+
+    mrmr = MRMRFeatureSelectionTransform(
+        relevance_table=relevance_table,
+        top_k=20,
+        model=CatBoostRegressor(iterations=1),
+        drop_zero=True,
```
Comment (d-a-bunin) [inline tests/test_transforms/test_feature_selection/test_feature_importance_transform.py:332]: Do we need this option here?

**Review by d-a-bunin — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 23]
```diff
@@ -873,4 +874,4 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - Distribution plot
   - Anomalies (Outliers) plot
   - Backtest (CrossValidation) plot
-  - Forecast plot
+  - Forecast plot
```
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply CHANGELOG.md:876]: Aren' you going to fix that?

File: tests/test_models/test_catboost.py [hunk 24]
```diff
@@ -200,7 +200,7 @@ def test_prediction_decomposition(ts_with_features, model):
     assert_prediction_components_are_present(model=model, train=train, test=test)
 
 
-@pytest.mark.parametrize("model", (CatBoostPerSegmentModel(), CatBoostMultiSegmentModel()))
+@pytest.mark.parametrize("model", (CatBoostPerSegmentModel(), CatBoostMultiSegmentModel(iterations=2000)))
```
Comment (d-a-bunin) [inline tests/test_models/test_catboost.py:203]: Why do we need to set `iterations=2000`?

**Review by d-a-bunin — APPROVED** [review]
